feat(api): TransactionValidator

MorganDream opened the pull request:
Transaction Validator for Neo3, can validate:
- validUntilBlock
- systemFee
- networkFee
- signatures

Validation will return ValidationResult for developer to handle.

MorganDream commented:
> For the validator, let me propose an alternative design. From a wallet's perspective, I imagine the usage to be:

> Craft transaction with initial user intents and initial fee estimates.
Perform validation and display the changes to the user.
User agrees and proceeds to signing.
To be inline with this, we should allow a codeflow that can surface the changes and then allow the system to accept the appropriate changes. Instead of rejecting the promises when there are changes, we should return the changes instead.

@snowypowers I quoted your suggestion on previous PR and let's discuss it here.
Validation of transaction is not only about fee, it can also validate intents and signatures. If intents and signatures is modified after validation, the previous validation for fees should be outdated. It's not applicable in this case to apply those changes.

Please check my new design. I suggest not to provide functions to apply validation suggestions, but just return it to developers and let developers handle this.

snowypowers commented:
1. It is weird that your individual functions return `ValidationResult` but only one of the fields will be filled.

2. Because signatures affect the cost of the transaction, I was considering adding unsigned witnesses during the builder phase so we can calculate the cost effectively.

3. I suggested a combined way in order to allow accepting suggestions. Im okay with developers handling the suggestions themselves.

4. I would like to be consistent and return a consistent result structure from all validation methods. It should contain a machine readable suggestion and also a human readable message.

MorganDream commented:
> 1. It is weird that your individual functions return `ValidationResult` but only one of the fields will be filled.

I haven't added the `validateAll` function, so user could do several validations at one time, then the result will contains several fields. Individual function returns the same type is easy for them to combine to one result.
> 2. Because signatures affect the cost of the transaction, I was considering adding unsigned witnesses during the builder phase so we can calculate the cost effectively.

Why not validate the transaction after user has signed it? or you mean to calculate in builder?
> 4. I would like to be consistent and return a consistent result structure from all validation methods. It should contain a machine readable suggestion and also a human readable message.

`ValidationResult` is a consistent result type. And it now has a machine readable message.
Well I agree to add a human readable message.

snowypowers commented:
> I haven't added the `validateAll` function, so user could do several validations at one time, then the result will contains several fields. Individual function returns the same type is easy for them to combine to one result.

How are you going to combine `result` and `fixed` into one result? On a side note, what does `result` even mean? I think it would be better to return `ValidationSuggestion` for each validation function if we are doing each one individually. How are you going to let the user control `validateAll`? 

> Why not validate the transaction after user has signed it? or you mean to calculate in builder?

Because at this stage there isnt much use of validation. Any changes to the transaction will cause a resign so might as well just send the transaction to the network.

If we add unsigned witnesses at the builder stage, we can calculate the signing fees correctly and produce quite a good fee estimate.

MorganDream commented:
> How are you going to combine result and fixed into one result? On a side note, what does result even mean? 

`result` means whether the transaction is valid in current validation.
`fixed` means whether the transaction has been modified by validator to make it valid. This can be TRUE when user specify `autoFix` as TRUE. But there are cases validator cannot fix like signing. `fixed` doesn't mean the transaction is fixed to be valid, but modified to achieve validation.
```ts
public async validateAll(autoFix = false) {
   return Promise.all([this.validateSystemFee()...]).then(res => {
     result: AND between all result,
     fixed: OR between all fixed
   })
}
```
Yeah, I can see the problem may be user don't know which validation is fixed which is not.
So how about moving fixed into validationSuggestion? Result can tell user whether transaction is valid after `validateAll`, `suggestions` can tell them the suggestion and what is fixed.

> Because at this stage there isnt much use of validation. Any changes to the transaction will cause a resign so might as well just send the transaction to the network.

> If we add unsigned witnesses at the builder stage, we can calculate the signing fees correctly and produce quite a good fee estimate.
Well as far as I can see, I think fee estimation is useless. As a user, I always want to spend as little as possible gas to complete my transaction, fee estimation will probably be invalid or make me pay more, which make it feel useless.

MorganDream commented:
A better and more scalable design may be that:
```ts
type ValidationAttr = 'systemFee' | 'networkFee' | 'witness';
interface ValidationSuggestion<T> {
  attr: ValidationAttr,
  prev: T,
  suggestion: T
}
public async validate(attrs: Array<ValidationAttr>, autoFix: boolean);

```

snowypowers commented:
Yes I would think that `ValidationResult` simply be `{ [transactionProperty: string]: ValidationSuggestion }`. I prefer using a map over an array as order wouldnt matter.

`result` field is kinda useless cos if we are giving a suggestion, it means its not valid.

Im okay with the `validate` design. Please use flags for the type, it will be nicer.

MorganDream commented:
> Yes I would think that ValidationResult simply be { [transactionProperty: string]: ValidationSuggestion }
There are some cases that we cannot fix and give the same type suggestion: Signing.
Or should we remove it?

> result field is kinda useless cos if we are giving a suggestion, it means its not valid.
If user set `autoFix: true` and validator fixes all to make it valid, we should give the ValidationSuggestion to tell them it's changed from before validation, and result is a fast property to tell them the transaction is valid now.

In newest code, I have change `result` to `valid`

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,229 @@
+   * validate systemFee
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  validateSystemFee = async (autoFix = false): Promise<ValidationResult> => {
```
1. You might want to do a check on the existing systemFee to make sure that it is in an acceptable format too (integer without decimals).

2. Consider including the action to verify the intent here too. If the intent is invalid, the systemFee validation will also probably be worthless.

MorganDream replied:
1. Yeah, thanks for reminding.
2. OK, will do. Planning to using manifest

snowypowers replied:
Im not clear if in neo3 the VM will error out on scripts requiring verification. If it does, then it might be hard to validate the script and estimate the fee.

MorganDream replied:
Not very clear about your concern, I have added `validateIntents` using ContractManifest.

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,229 @@
+      };
+    }
+    return {
+      valid: true
```
So there isnt a suggestion if we do not want autofix but attached a larger than needed fee?

MorganDream replied:
Yes. Anyway validator is mostly to validate, to save money is secondary job. But it's harmless if we add a check here, what do you think?

snowypowers replied:
If you say that, then autofix should not fix if the attached fee is larger than required.

MorganDream replied:
Well I think autoFix could fix when attached fee is larger than required.
It will not give suggestions when autoFix = false and attached fee is larger than required.
What do you think?

snowypowers replied:
I think most people will expect the `autoFix` flag to just apply the suggestion and not change the suggestion output based on that flag. Please make it consistent.

MorganDream replied:
OK.

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,229 @@
+    };
+  };
+
+  // validateIntents = async (): Promise<ValidationResult> => {
```
Remove comment

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,229 @@
+  suggestion: T;
+}
+
+export interface ValidationResult {
```
Consider 
```
{ 
  valid: boolean,
  suggestions: {[prop: string] : ValidationSuggestion<T>
} where prop in Transaction
```

If there is no way to suggest a fix, leave suggestion `null` but offer a human readable message on the issue.

MorganDream replied:
Cannot use `[prop: string] : ValidationSuggestion<T>` cause we need to pre-define `T`.
now it's 
```ts
suggestions?: {
    validaUntilBlock?: ValidationSuggestion<number>;
    systemFee?: ValidationSuggestion<u.Fixed8>;
    networkFee?: ValidationSuggestion<u.Fixed8>;
    witnesses?: string;
  };
```

snowypowers replied:
Can we have the suggestion reflect the properties of Transaction? So instead of `witnesses` its `scripts`.

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,229 @@
+
+  async validate(
+    attrs: ValidationAttributes,
+    autoFix = false
```
can we also have `autoFix` be `ValidationAttributes` so we can select properties to autofix?

MorganDream replied:
Emm.. then the flag thing should be changed, right?
I believe user can invoke twice, one with autoFix, one without to achieve this scope.

snowypowers replied:
> Emm.. then the flag thing should be changed, right?

What do you mean? 

> I believe user can invoke twice, one with autoFix, one without to achieve this scope

Yes they could. I think the autoFix change is a small natural improvement using the same set of flags.

MorganDream replied:
Oh, I see it. Will do.

MorganDream replied:
`autoFix: ValidationAttributes | boolean`
Give user the freedom to assign one boolean, too.

snowypowers replied:
you can create the convenient flags `ValidationAttributes.ALL` and  `ValidationAttributes.NONE`

MorganDream commented on packages/neon-api/src/transaction/util.ts:
```diff
@@ -0,0 +1,64 @@
+  const signers = transaction.getScriptHashesForVerifying();
+  signers.forEach(signer => {
+    const account = new wallet.Account(signer);
+    if (!account.isMultiSig) {
```
Now I calculate fee from signatures

MorganDream commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,229 @@
+  /**
+   * Validate Signatures
+   */
+  validateSigning = async (): Promise<ValidationResult> => {
```
Signatures are part of transaction and signing should be done before validation I think.

snowypowers replied:
I thought we have discussed this multiple times. The order should be build, validate and sign. Signing must always be the last step because any adjustments during the build and validate step will require a resigning.

MorganDream replied:
I have removed validateSigning, however how can we add such empty witnesses?

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,298 @@
+      }
+    }
+
+    return undefined;
```
You might want to consider returning empty string or null instead of undefined. Undefined is usually not intentional.

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,298 @@
+      return {
+        valid: false,
+        suggestions: {
+          script: intentsRes
```
We should return a ValidationSuggestion and add a `message` field that indicates a human readable error. At this point, you might want to not add the `message` feature first and complete this PR.

MorganDream replied:
I just added this. Please check

snowypowers commented on packages/neon-core/src/rpc/RPCClient.ts:
```diff
@@ -220,10 +221,14 @@ export class RPCClient {
+  ): Promise<ContractManifest | null> {
     const response = await this.execute(Query.getContractState(scriptHash));
-    return response.result;
+    if (response.error) {
```
the rpc client will throw if there is an error. I dont think hiding the error is a good idea anyway.

MorganDream replied:
If we just catch the throwing error, we cannot tell exactly whether the error is because of network error or contract not exist.

If contract doesn't exist, it will return 200, too, with an error info in the response body. We can firmly tell user it doesn't exist.

snowypowers replied:
Ah i see my bad I misread the code. Maybe time to rework the RPC client a little.

MorganDream commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,303 @@
+  /**
+   * Validate intents
+   */
+  async validateScript(): Promise<ValidationResult> {
```
Indeed.
Something that need to be considered: 
1. If we add online testing for script in validateScript and user validate both systemFee and script, the same request would be sent twice.
2. At the begging of testing on neo3 preview testnet, a script will return 'FAULT' when `invokeScript` but actually works in a transaction. I'll try to reproduce this problem and ask someone else if it is a bug.

snowypowers replied:
in my experience in neo2, fault can happen because the contract method requires a signature check which an invokescript is unable to provide. However, the script is valid.

MorganDream replied:
So shall we use `invokeScript` to validate script? Considering it may return `FAULT` in a valid script.

snowypowers replied:
If it fails, we probably set a warning stating that it produced a fault. Autofix should also fail since we cannot accurately estimate the cost if it fails. We could just give the GAS used up to the failure as a baseline in the message.

MorganDream commented:
I'll add test later.

MorganDream commented:
@snowypowers Now ut is added

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,314 @@
+      fixed?: boolean;
+      prev?: T;
+      suggestion?: T;
+      message: T;
```
`message` field should be string.

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,314 @@
+      suggestion: T;
+      message?: T;
+    }
+  | {
```
Is there any way we can simplify this? Do we have to make `fixed` optional or can we just always return something (`false` if `autoFix` is false?).

snowypowers commented on packages/neon-api/__tests__/transaction/validator.ts:
```diff
@@ -0,0 +1,320 @@
+    const transaction = new TransactionBuilder({
+      validUntilBlock:
+        tx.Transaction.MAX_TRANSACTION_LIFESPAN +
+        (await rpcClient.getBlockCount()) -
```
I think we can move the entire suite to the integration folder instead.

snowypowers commented on packages/neon-api/__tests__/transaction/validator.ts:
```diff
@@ -0,0 +1,320 @@
+    }).build();
+    const validator = new TransactionValidator(rpcClient, transaction);
+    const result = await validator.validateValidUntilBlock();
+    expect(result.valid).toBeFalsy();
```
Not checking the other fields?

MorganDream replied:
Check the whole object when autoFix = true. Here we only care about whether it's valid

snowypowers commented on packages/neon-api/__tests__/transaction/validator.ts:
```diff
@@ -0,0 +1,320 @@
+    expect(result.valid).toBeTruthy();
+  });
+
+  test("inValid", async () => {
```
typo

snowypowers commented on packages/neon-api/__tests__/transaction/validator.ts:
```diff
@@ -0,0 +1,320 @@
+    expect(result.valid).toBeTruthy();
+  });
+
+  test("inValid", async () => {
```
typo

snowypowers commented on packages/neon-api/__tests__/transaction/validator.ts:
```diff
@@ -0,0 +1,320 @@
+    expect(result.valid).toBeFalsy();
+  });
+
+  test("autoFix", async () => {
```
add test for validating all but autofixing only 1.

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,314 @@
+      script
+    );
+    const minimumSystemFee = new u.Fixed8(
+      parseFloat(gasConsumed) * 1e-8
```
please dont do multiply outside of Fixed8.

MorganDream replied:
Why? If I use `multiplyBy` it will return `BigNumber` and cannot use `ceil` after that.

snowypowers replied:
doing multiplication in js introduces inaccuracies. There is `times` or `mul`. have you tried that?

MorganDream replied:
OK

MorganDream commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,337 @@
+  public async validateSystemFee(autoFix = false): Promise<ValidationResult> {
+    const validateIntentsResult = await this.validateScript();
+    if (!validateIntentsResult.valid) {
+      return validateIntentsResult;
```
I removed validating script here.

snowypowers commented on packages/neon-api/__integration__/transaction/validator.ts:
```diff
@@ -0,0 +1,388 @@
+      }).build();
+      const validator = new TransactionValidator(rpcClient, transaction);
+      const result = await validator.validateSystemFee();
+      expect(result.valid).toBeFalsy();
```
should there be an expectation of a suggestion?

MorganDream replied:
The suggested value is dynamically changing as it's related to block height.

snowypowers replied:
You can check for the existence of the suggestion and it be bigger than the current number.

snowypowers commented on packages/neon-api/__integration__/transaction/validator.ts:
```diff
@@ -0,0 +1,388 @@
+      }).build();
+      const validator = new TransactionValidator(rpcClient, transaction);
+      const result = await validator.validateSystemFee();
+      expect(result.valid).toBeFalsy();
```
Similarly, there should be a suggestion here?

MorganDream replied:
OK

snowypowers commented on packages/neon-api/__integration__/transaction/validator.ts:
```diff
@@ -0,0 +1,388 @@
+    }).build();
+    const validator = new TransactionValidator(rpcClient, transaction);
+    const result = await validator.validateValidUntilBlock();
+    expect(result.valid).toBeFalsy();
```
suggestion?

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,344 @@
+              fixed: false,
+              message: "Encountered FAULT when validating script."
+            },
+            ...result.suggestions?.systemFee
```
You might want to change this such that the suggestion only appears if the current attached fee is lower than what was consumed before it faulted.

MorganDream replied:
Message changed

snowypowers commented on packages/neon-api/src/transaction/validator.ts:
```diff
@@ -0,0 +1,345 @@
+              message:
+                "Cannot get precise systemFee as script execution on node reports FAULT."
+            },
+            ...validation.result?.systemFee
```
I thought we discussed and this suggestion should only appear when the suggestion is smaller than the currently attached fee.

hal0x2328 commented:
500 points MorganDream
300 points snowypowers for review

lllwvlvwlll commented:
https://neoscan.io/transaction/44C84C9E7582DC00411BC9D9D075F6993BA7E203FFDDE366E21DFEBD125EA580
